pkg/kubectl/cmd/util/factory_client_access.go

```diff
@@ -50,137 +46,48 @@ import (
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
-	utilflag "k8s.io/apiserver/pkg/util/flag"
 	"k8s.io/client-go/discovery"
 	"k8s.io/client-go/kubernetes"
 	restclient "k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/clientcmd"
-	"k8s.io/client-go/util/homedir"
 	"k8s.io/kubernetes/pkg/api/legacyscheme"
 	"k8s.io/kubernetes/pkg/apis/apps"
 	api "k8s.io/kubernetes/pkg/apis/core"
 	"k8s.io/kubernetes/pkg/apis/extensions"
 	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
 	"k8s.io/kubernetes/pkg/kubectl"
 	"k8s.io/kubernetes/pkg/kubectl/resource"
-	"k8s.io/kubernetes/pkg/kubectl/util/transport"
 	"k8s.io/kubernetes/pkg/version"
 )
 
+type RESTClientGetter interface {
+	ToRESTConfig() (*restclient.Config, error)
+	ToDiscoveryClient() (discovery.CachedDiscoveryInterface, error)
+	ToRawKubeConfigLoader() clientcmd.ClientConfig
+}
+
 type ring0Factory struct {
-	flags             *pflag.FlagSet
-	clientConfig      clientcmd.ClientConfig
-	discoveryCacheDir string
+	clientGetter RESTClientGetter
 
 	requireMatchedServerVersion bool
 	checkServerVersion          sync.Once
 	matchesServerVersionErr     error
 }
 
-func NewClientAccessFactory(optionalClientConfig clientcmd.ClientConfig) ClientAccessFactory {
-	flags := pflag.NewFlagSet("", pflag.ContinueOnError)
-
-	clientConfig := optionalClientConfig
-	if optionalClientConfig == nil {
-		clientConfig = DefaultClientConfig(flags)
+func NewClientAccessFactory(clientGetter RESTClientGetter) ClientAccessFactory {
+	if clientGetter == nil {
+		panic("attempt to instantiate client_access_factory with nil clientGetter")
```

Comment:
I'm only going to take this because it's a coding error to pass this.  In general, a construct like this is ugly.